File: leetcode/path_sum_II/path_sum_II.cpp
/*
 * Time Complexity: O(N)
 *   - Each node is visited once in the DFS traversal.
 *
 * Space Complexity: O(H)
 *   - H is the height of the tree (O(log N) for balanced tree, O(N) for skewed tree).
 *   - Due to recursion stack and the path vector.
 */

#include <vector>

using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    vector<vector<int>> ans;

    void dfs(TreeNode* root, vector<int>& currNodes, int currSum, int targetSum) {
        currNodes.push_back(root->val);
        currSum += root->val;

        if(root->left == nullptr && root->right == nullptr) {
            if(currSum == targetSum) {
                ans.push_back(currNodes);
            }
        } else {
            if(root->left) dfs(root->left, currNodes, currSum, targetSum);
            if(root->right) dfs(root->right, currNodes, currSum, targetSum);
        }

        currNodes.pop_back();
    }

public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        if(root == nullptr) return ans;
        vector<int> currNodes;
        dfs(root, currNodes, 0, targetSum);
        return ans;
    }
};

File: leetcode/path_sum_II/path_sum_II.java
package path_sum_II;

import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private List<List<Integer>> ans = new ArrayList<>();

    private void dfs(TreeNode root, List<Integer> currNodes, int currSum, int targetSum) {
        currNodes.add(root.val);
        currSum += root.val;

        if (root.left == null && root.right == null) {
            if (currSum == targetSum) {
                ans.add(new ArrayList<>(currNodes));
            }
        } else {
            if (root.left != null) dfs(root.left, currNodes, currSum, targetSum);
            if (root.right != null) dfs(root.right, currNodes, currSum, targetSum);
        }

        currNodes.remove(currNodes.size() - 1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) return ans;
        dfs(root, new ArrayList<>(), 0, targetSum);
        return ans;
    }
}
